Refs #24840 -- Added GDALRaster warp and transform methods.

yellowcap opened the pull request:
- The warp method allows altering any parameters of the raster (geotranform, datatype, srid, etc
- The transform method allows reprojecting the raster into a different srid

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -278,3 +280,122 @@ def bands(self):
+        input parameter. The default is NearestNeighbor. For a list of all options
+        consult the GDAL_RESAMPLE_ALGORITHMS constant.
+        """
+        # Get the parameters defining the geotransform, srid and size of the raster
```
, and

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -278,3 +280,122 @@ def bands(self):
+        if 'skew' not in ds_input:
+            ds_input['skew'] = self.skew
+
+        # Get the driver, name and datatype of the target raster
```
, and

timgraham commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -278,3 +280,122 @@ def bands(self):
+        data = {
+            'origin': [origin.x, origin.y],
+            'scale': [scalex, scaley],
+            'srid': srid
```
srid,

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+
+        Returns a warped version of this raster.
+        
+        The warping parameters can be specified through the `ds_input` argument.
```
Need double backticks to have an effect on the rendered docs.

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+        origin, scale, skew, datatype, driver, and name (filename).
+
+        By default, the warp functions keeps most parameters equal to the values
+        of the original source raster. So only parameters that should be changed
```
, so only...

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+
+        By default, the warp functions keeps most parameters equal to the values
+        of the original source raster. So only parameters that should be changed
+        need to be specified. Note that this includes the driver, so for file based
```
file-based

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+        
+        The only parameter that is set differently from the source raster is the
+        name. The default value of the the raster name is the name of the source
+        raster appended with `_copy + source_driver_name`. For file based rasters
```
'_copy' + source_driver_name

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+        raster appended with `_copy + source_driver_name`. For file based rasters
+        it is recommended to provide the file path of the target raster.
+
+        The re-sampling algorithm used for warping can be specified with the
```
resampling

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+        it is recommended to provide the file path of the target raster.
+
+        The re-sampling algorithm used for warping can be specified with the
+        `resampling` argument. The default is `NearestNeighbor`, the other
```
, and the other

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+            ...     "width": 6, "height": 6, "srid": 3086,
+            ...     "origin": [500000, 400000],
+            ...     "scale": [100, -100],
+            ...     "bands": [{"data": range(36), "nodata_value": 99}]
```
trailing comma

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+                   [ 19.,  21.,  23.],
+                   [ 31.,  33.,  35.]], dtype=float32)
+
+    .. method:: transform(srid, driver=None, name=None, squared_pixels=True, resampling='NearestNeighbour'):
```
chop ending colon

timgraham commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1288,6 +1288,83 @@ blue.
+
+        By default, the `x` and `y` scales of the pixels of the transformed raster
+        are set to be equal. The larger of the two values in the new projection is
+        used. To disable this behavior set the `squared_pixels` property to `False`.
```
behavior,

timgraham commented on tests/gis_tests/gdal_tests/test_raster.py:
```diff
@@ -155,6 +155,101 @@ def test_file_based_raster_creation(self):
+            'skew': (0, 0),
+            'bands': [{
+                'data': range(16),
+                'nodata_value': 255
```
trailing comma here and next line

timgraham commented:
Add 1.9 release notes?

yellowcap commented:
Thanks for the review @timgraham . I added 1.9 release notes and included the corrections. However, Jenkins does not seem to like the word `resampling`, is there a way to add it to the spell checker? Otherwise we can just use `sampling` which would also work in this context.

timgraham commented:
Add it to docs/spelling_wordlist.

yellowcap commented:
I corrected the spelling issue and I rewrote the ``transform`` method to use the ``GDALAutoCreateWarpedVRT`` function in gdal. The result is very similar, but I think this method is more stable as it uses gdal to compute the transformed geotransform parameters. I think like this it will be more reliable overall.

timgraham commented:
Some cosmetic and doc edits: http://dpaste.com/163BPM4
LGTM, pending @claudep's review of the functionality.

claudep commented:
Great, committed in c0780215
